File: apps/flutter_wallet/lib/src/core/native/parsers/transfers.dart
import '../../../models.dart';
import '../../../wallet_api.dart';
import 'fields.dart';

TransferPreview transferPreviewFromNativeJson(Object? value) {
  if (value is! Map<String, dynamic>) {
    throw const WalletApiException('Invalid native response');
  }
  return TransferPreview(
    chain: nativeStringField(value, 'chain'),
    fromAddress: nativeStringField(value, 'from_address'),
    toAddress: nativeStringField(value, 'to_address'),
    assetSymbol: nativeStringField(value, 'asset_symbol'),
    amount: nativeStringField(value, 'amount'),
    feeEstimate: nativeStringField(value, 'fee_estimate'),
    rpcUrl: nativeStringField(value, 'rpc_url'),
    resourceStatus: transferResourceStatusFromNativeJson(
      value['resource_status'],
    ),
  );
}

TransferResourceStatus? transferResourceStatusFromNativeJson(Object? value) {
  if (value == null) {
    return null;
  }
  if (value is! Map<String, dynamic>) {
    throw const WalletApiException('Invalid native response');
  }
  return TransferResourceStatus(
    energyAvailable: nativeIntField(value, 'energy_available'),
    energyRequired: nativeIntField(value, 'energy_required'),
    bandwidthAvailable: nativeIntField(value, 'bandwidth_available'),
    bandwidthRequired: nativeIntField(value, 'bandwidth_required'),
    trxBalanceSun: nativeIntField(value, 'trx_balance_sun'),
    trxFeeReserveRequiredSun: nativeIntField(
      value,
      'trx_fee_reserve_required_sun',
    ),
    canSendWithoutBurningTrx: nativeBoolField(
      value,
      'can_send_without_burning_trx',
    ),
  );
}

TransferResult transferResultFromNativeJson(Object? value) {
  if (value is! Map<String, dynamic>) {
    throw const WalletApiException('Invalid native response');
  }
  return TransferResult(
    chain: nativeStringField(value, 'chain'),
    txHash: nativeStringField(value, 'tx_hash'),
    status: nativeStringField(value, 'status'),
  );
}

ParsedPayment parsedPaymentFromNativeResponse(Object? body) {
  if (body is! Map<String, dynamic>) {
    throw const WalletApiException('Invalid native response');
  }
  final address = body['address'];
  if (address is! String) {
    throw const WalletApiException('Invalid native response');
  }
  return ParsedPayment(
    address: address,
    chain: optionalNativeString(body, 'chain'),
    amount: optionalNativeString(body, 'amount'),
    note: optionalNativeString(body, 'note'),
  );
}

File: crates/wallet_core/src/models/transfers.rs
use super::ChainId;
use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct TransferRequest {
    pub wallet_id: Uuid,
    pub chain: ChainId,
    pub asset_id: Uuid,
    pub to_address: String,
    pub amount: String,
    #[serde(default)]
    pub block_if_energy_insufficient: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct TransferPreview {
    pub chain: ChainId,
    pub from_address: String,
    pub to_address: String,
    pub asset_symbol: String,
    pub amount: String,
    pub fee_estimate: String,
    pub rpc_url: String,
    pub resource_status: Option<TransferResourceStatus>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct TransferResourceStatus {
    pub energy_available: u64,
    pub energy_required: u64,
    pub bandwidth_available: u64,
    pub trx_balance_sun: u64,
    pub trx_fee_reserve_required_sun: u64,
    pub can_send_without_burning_trx: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct TransferResult {
    pub chain: ChainId,
    pub tx_hash: String,
    pub status: String,
}
